### ISSUE TITLE
The `be_reachable` matcher in the Host resource should use the port/protocol provided

### ISSUE THREAD

**Issue by cullenmcdermott** [inspec/inspec#1439]
### Description

The `reachable?` method for the Host resource should do more than check if a host is reachable via ping. If ping is blocked then the test will fail.

### InSpec and Platform Version

Inspec 1.10.0
Ubuntu 16.04

### Replication Case

1) Setup a firewall rule that blocks icmp traffic from Host A to Host B
2) Create an simple test that checks reachability on Host A and run it:
```
describe host("#{Host_A}", port: 22, proto: 'tcp') do
    it { should be_reachable }
  end
```

### Expected Outcome

If I am specifying a port and protocol I expect that to be what is tested as sometimes icmp traffic is blocked by a firewall. 

### Possible Solutions

A possible solution would be to execute nc on the host running the test. However not all distributions come with nc installed so that could pose an issue.

A better solution might be to open a tcp(or udp) socket with bash.

The best solution would probably be to use [socket](https://ruby-doc.org/stdlib-2.3.1/libdoc/socket/rdoc/Socket.html) since that should be platform agnostic and work even if bash isn't available.

**Comment by MooreDerek** [comment]
I'd be interested in helping to fix this, from the proposed solutions there are some interesting issues when looking at Windows and *nix support of available commands.

**Proposed Solution - \*nix**

Use, in order, *nc, wget, telnet, bash* then *ping*.  Use *exists* to determine if they are installed and report in the output which command was used.  

The main issue that I've come across in testing is the timeout on both *telnet* and *bash* when testing a port with no listener; *telnet* timeouts after 60 seconds and *bash* can take over 10 minutes.

*wget* and *telnet* do not provide udp support, but as it's not guaranteed delivery, not sure that's a big issue.

**Proposed Solution - Windows**

Use, in order, *Test-NetConnection, telnet* and then *ping*, using exists to verify installation and report in the output.

Main issue is *Test-NetConnection* is only available in Windows 8.1 and upwards, and *telnet* tends to be removed on a lot of enterprise Windows installs. Again timeout on *telnet* is a consideration.

Thoughts?

**Comment by adamleff** [comment]
@MooreDerek I'm super excited you're willing to contribute to this, thank you!

Some of my thoughts:

 * For tools where we cannot control the timeout, I'd simply eliminate them as possible options (i.e. telnet and bash) until we can add functionality to Train for the caller to provide an execution timeout parameter and have it handle that gracefully with proper exception raising. Currently, to my knowledge, that doesn't exist... so even if we wrapped our command calls in InSpec with `Timeout.timeout`, we'd still be blocking execution... no bueno!

 * Since wget and telnet do not provide UDP support, I wouldn't consider them as options if the user supplies `udp` as the protocol. I think it would be rather easy for us to exclude those for udp requests.

 * For the Windows stuff, that sounds fine. If the commands aren't available, we don't use them. :) Again, if we can't control the telnet timeout, I'd eliminate that from consideration for now.

**Comment by jkugler** [comment]
Why not use socket, as was proposed? Then you don't have Unix/Windows branching, and have native-language tools (don't have to shell out).

**Comment by adamleff** [comment]
@jkugler InSpec has to work for local executions as well as remote. We have no guarantee that the remote system will have Ruby on it, nor does InSpec install anything in order to perform its scans. Doing an inline script to ruby using socket may be an option, but like all the other proposed options, it may not be available.

**Comment by jkugler** [comment]
The "is reachable" checks if it's reachable from the machine invoking the test (the machine running bin/inspec), no?  So it would have everything available to it that Ruby has?
```
# matcher to check if host is reachable
RSpec::Matchers.define :be_reachable do
  match do |host|
    host.reachable? == true
  end
```
The question of whether or not it is reachable is answered in ruby code, yes? From docs
```
### Verify host name is reachable over a specific protocol and port number

    describe host('example.com', port: 53, proto: 'udp') do
      it { should be_reachable }
    end
```
reachable from where? From the machine running bin/inspec, presumably, which is a ruby script.

Also, I see this:
```
  class LinuxHostProvider < HostProvider
    # ping is difficult to achieve, since we are not sure
    def ping(hostname, _port = nil, _proto = nil)
      # fall back to ping, but we can only test ICMP packages with ping
      # therefore we have to skip the test, if we do not have everything on the node to run the test
      ping = inspec.command("ping -w 1 -c 1 #{hostname}")
      ping.exit_status.to_i != 0 ? false : true
    end
...
end
```
That is executing the shell out on the box on which `bin/inspec` is running. Why are we not able to load up Ruby's socket right there instead?

**Comment by rockandska** [comment]
Hi @adamleff , @jquick , @jvale ,

Despite the merge #2607 in v2.0.45 (2018-03-08), it seems that host resource don't use bash over netcat to test host even if bash and timeout are available or did i missed something ?


```
control 'test-mongodb-network' do
        title 'Test mongodb network'
        desc 'Check than we can reach the host/port defined in configuration'

        describe host( conf_file.DATABASE.host, port: conf_file.DATABASE.port, protocol: 'tcp') do
                it { should be_reachable }
        end
end
```
```
$ inspec --version
2.1.0

$ inspec exec . -t docker://bffcc19fe067cd78352ff0f7e46deb63cb1bac3f9b2f159ab43bc75b4a390289 --controls test-mongodb-network

Profile: tests from . (tests from .)
Version: (not specified)
Target:  docker://bffcc19fe067cd78352ff0f7e46deb63cb1bac3f9b2f159ab43bc75b4a390289

  ↺  test-mongodb-network: Test mongodb network
     ↺  The following requirements are not met for this resource: netcat must be installed


Profile Summary: 0 successful controls, 0 control failures, 1 control skipped
Test Summary: 0 successful, 0 failures, 1 skipped
```
```
$ docker exec -ti -u root bffcc19fe067cd78352ff0f7e46deb63cb1bac3f9b2f159ab43bc75b4a390289 /bin/bash
[root@bffcc19fe067]# which bash
/bin/bash
[root@bffcc19fe067 ]# which timeout
/usr/bin/timeout
[root@bffcc19fe067 ]# 
```

Additionally, is it a common way to issue a "test skipping" in place of a "test failure" if requirements are not satisfied ? 
Is there any possibility to force this test to be a "failure" if requirements are not satisfied ?

Thanks,

Regards, :)

**Comment by jvale** [comment]
@rockandska Regarding using bash or not, `timeout` being available isn't the only requirement, bash itself must have been compiled with `--enable-net-redirections` to be able to use `/dev/tcp` and friends. Is it possible that it isn't the case? Some distributions don't use that flag (read somewhere about Debian and Ubuntu, but never actually checked).

**Comment by ferrarimarco** [comment]
bump to avoid closing.

This is still an issue :(

**Comment by kekaichinose** [comment]
Thank you for tracking down this issue. As with many large open source projects, we have a backlog of bugs and issues of varying user impact. Some issues are perpetually subject to lower priority than others in our backlog. We have reviewed the issue, and we have determined this a very low priority when considered in the context of the entire backlog. We have elected to close it as "Declined" due to competing priorities. We very much appreciate you bringing this to our attention.

Thank you!
__Chef InSpec Team__

**Comment by sbonds** [comment]
Shoot-- this seems like an EXTREMELY common item to need to test as part of an environment build out. Being unable to reach specific TCP ports from specific sources is a major source of delays and troubleshooting. Knowing that these basics work well could be a real time-saver, so it's disappointing to see this feature treated as a low priority.

However, this is Open Source. Since Chef won't do this, has anyone made a third party module to do remote port checks?
